### PR TITLE
do-request: try every URL from repository JSON, before using fallback

### PR DESCRIPTION
**Pull request by josephfrazier**
https://www.npmjs.com/package/github-url-from-username-repo is an
example of a package that has a github homepage, but wasn't being
resolved correctly, due to some strange formatting of the URLs in the
JSON from registry.npmjs.org. This PR makes it so that every URL from
the JSON is tried (in order), before falling back to the npmjs.com URL.

EDIT: oops, I accidentally pushed this to a misnamed branch. Hope that's not an issue.

### CODE REVIEW

File: src/utils/do-request.js [hunk 1]
```diff
@@ -39,28 +39,32 @@ module.exports = async function doRequest(packageName, type) {
...
 
-    try {
-      await got.get(url);
-      return url;
-    } catch (err) {
-      url = util.format(config.fallback, packageName);
-      return url;
-    }
+    // If we get here, no urls could be loaded.
+    return util.format(config.fallback, packageName);
```
Comment (stefanbuck) [inline src/utils/do-request.js]: I think with [find-reachable-urls](https://github.com/stefanbuck/find-reachable-urls) we can simplify a lot here and get rid of the loop / await ugliness here.

Comment (josephfrazier) [inline src/utils/do-request.js]: I like that idea, but I'm having some trouble getting the test stubs set up properly. See https://github.com/OctoLinker/live-resolver/pull/34/commits/25d4763a7995f93a536923c163f48159c5c43802

Comment (stefanbuck) [inline src/utils/do-request.js]: The problem is that you can't stub a dependency that requires another dependency.

An easy option would be to mock `find-reachable-urls`or using [proxyquire](https://www.npmjs.com/package/proxyquire).

Comment (stefanbuck) [inline src/utils/do-request.js]: Ideally we could mock the request on the nodejs core level. By that we would be independent from the packages that we use.

Comment (stefanbuck) [inline src/utils/do-request.js]: [nock](https://github.com/node-nock/nock) looks pretty cool and powerful. The set up maybe needs some time. I would defiantly prefer `nock` over `proxyquire`, but for now we can go with mocking `find-reachable-urls` to keep it simple.

Comment (josephfrazier) [inline src/utils/do-request.js]: I like the idea of using `nock`, but I wasn't able to get it to work (see https://github.com/OctoLinker/live-resolver/pull/34/commits/ee9dfd3e6969c605efec0ee8effa6c5c0c0f0c96). It looks like mocking `find-reachable-urls` is rather difficult, so how about we just remove the failing tests, and add a functional test to ensure the relevant behavior is still covered? See https://github.com/OctoLinker/live-resolver/pull/34/commits/457bb62a2d78329ee23114f32e200e40a477978b

Comment (stefanbuck) [inline src/utils/do-request.js]: Fine for me 👍

**Comment by stefanbuck** [comment]
Great improvement. Thanks @josephfrazier

**Comment by josephfrazier** [comment]
Thanks!

**Pull request merged**